File: android/src/shared/com/smileidentity/shared/views/SmileIDEnhancedDocumentVerificationView.kt
package com.smileidentity.shared.views

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.lifecycle.viewmodel.compose.LocalViewModelStoreOwner
import com.smileidentity.SmileID
import com.smileidentity.compose.EnhancedDocumentVerificationScreen
import com.smileidentity.react.results.DocumentCaptureResult
import com.smileidentity.react.utils.DocumentCaptureResultAdapter
import com.smileidentity.results.SmileIDResult
import com.smileidentity.shared.SmileIDSharedResult
import com.smileidentity.shared.SmileIDViewConfig
import com.smileidentity.util.randomJobId
import com.smileidentity.util.randomUserId

@Composable
fun SmileIDEnhancedDocumentVerificationView(
  config: SmileIDViewConfig,
  onResult: (SmileIDSharedResult<*>) -> Unit
) {
  config.countryCode ?: run {
    onResult(SmileIDSharedResult.WithError(IllegalArgumentException("countryCode is required for EnhancedDocumentVerification")))
    return
  }

  config.consentInformation ?: run {
    onResult(SmileIDSharedResult.WithError(IllegalArgumentException("consentInformation is required for EnhancedDocumentVerification")))
    return
  }

  val customViewModelStoreOwner = CustomViewModelStoreOwner()
  CompositionLocalProvider(LocalViewModelStoreOwner provides customViewModelStoreOwner) {
    SmileID.EnhancedDocumentVerificationScreen(
      userId = config.userId ?: rememberSaveable { randomUserId() },
      jobId = config.jobId ?: rememberSaveable { randomJobId() },
      countryCode = config.countryCode!!,
      documentType = config.documentType,
      idAspectRatio = config.idAspectRatio,
      showAttribution = config.showAttribution,
      allowAgentMode = config.allowAgentMode,
      showInstructions = config.showInstructions,
      allowNewEnroll = config.allowNewEnroll,
      allowGalleryUpload = config.allowGalleryUpload,
      captureBothSides = config.captureBothSides,
      extraPartnerParams = config.extraPartnerParams,
      consentInformation = config.consentInformation!!,
      useStrictMode = config.useStrictMode,
      ) { res ->
        when (res) {
          is SmileIDResult.Success -> {
            val result =
              DocumentCaptureResult(
                selfieFile = res.data.selfieFile,
                documentFrontFile = res.data.documentFrontFile,
                livenessFiles = res.data.livenessFiles,
                documentBackFile = res.data.documentBackFile,
                didSubmitEnhancedDocVJob = res.data.didSubmitEnhancedDocVJob,
              )
            val newMoshi =
              SmileID.moshi
                .newBuilder()
                .add(DocumentCaptureResultAdapter.FACTORY)
                .build()
          val json =
            try {
              newMoshi
                .adapter(DocumentCaptureResult::class.java)
                .toJson(result)
            } catch (e: Exception) {
              onResult(SmileIDSharedResult.WithError(e))
              return@EnhancedDocumentVerificationScreen
            }
          json?.let { js ->
            onResult(SmileIDSharedResult.Success(js))
          }
        }

        is SmileIDResult.Error -> onResult(SmileIDSharedResult.WithError(res.throwable))
      }
    }
  }
}

File: package/ios/Shared/Views/SmileIDEnhancedDocumentVerificationView.swift
//
//  SmileIDEnhancedDocumentVerificationView.swift
//  SmileIDShared
//
//  Created by Japhet Ndhlovu on 6/19/25.
//

import Foundation
import SmileID
import SwiftUI

/// Shared EnhancedDocumentVerification view that uses SmileIDViewConfig
@available(iOS 13.0, *)
public struct SmileIDEnhancedDocumentVerificationView: View {
    let config: SmileIDViewConfig
    let onResult: (SmileIDSharedResult<Any>) -> Void
    
    public init(
        config: SmileIDViewConfig,
        onResult: @escaping (SmileIDSharedResult<Any>) -> Void
    ) {
        self.config = config
        self.onResult = onResult
    }
    
    public var body: some View {
        Group {
            // Use SmileIDViewConfig's built-in validation
            if !config.validateEnhancedDocumentProperties() {
                // Return error immediately if validation fails
                Text("")
                    .onAppear {
                        let error = NSError(
                            domain: "SmileIDEnhancedDocumentVerificationView",
                            code: -1,
                            userInfo: [NSLocalizedDescriptionKey: "Required properties for EnhancedDocumentVerification are missing"]
                        )
                        onResult(.withError(cause: error))
                    }
            } else {
                // Configuration is valid, show the EnhancedDocumentVerification screen
                EnhancedDocumentVerificationContent(
                    config: config,
                    onResult: onResult
                )
            }
        }
    }
}

@available(iOS 13.0, *)
private struct EnhancedDocumentVerificationContent: View {
    let config: SmileIDViewConfig
    let onResult: (SmileIDSharedResult<Any>) -> Void
    
    var body: some View {
        SmileID.enhancedDocumentVerificationScreen(
            countryCode: config.countryCode!,
            documentType: config.documentType,
            userId: config.userId ?? SmileID.randomUserId(),
            jobId: config.jobId ?? SmileID.randomJobId(),
            allowNewEnroll: config.allowNewEnroll,
            allowAgentMode: config.allowAgentMode,
            showAttribution: config.showAttribution,
            allowGallerySelection: config.allowGalleryUpload,
            showInstructions: config.showInstructions,
            idAspectRatio: config.idAspectRatio,
            captureBothSides: config.captureBothSides,
            skipApiSubmission: config.skipApiSubmission,
            useStrictMode: config.useStrictMode,
            extraPartnerParams: config.extraPartnerParams,
            consentInformation: config.consentInformation!,
            delegate: EnhancedDocumentVerificationResultHandler(onResult: onResult)
        )
    }
}

/// Result handler that converts SDK results to shared results
private class EnhancedDocumentVerificationResultHandler: EnhancedDocumentVerificationResultDelegate {
    let onResult: (SmileIDSharedResult<Any>) -> Void
    
    init(onResult: @escaping (SmileIDSharedResult<Any>) -> Void) {
        self.onResult = onResult
    }
    
    func didSucceed(selfieImage: URL, documentFrontImage: URL, documentBackImage: URL?, didSubmitEnhancedDocVJob: Bool) {
        let result: [String: Any] = [
            "selfieFile": selfieImage.path,
            "documentFrontFile": documentFrontImage.path,
            "documentBackFile": documentBackImage?.path as Any,
            "didSubmitEnhancedDocVJob": didSubmitEnhancedDocVJob
        ]
        onResult(.success(result))
    }
    
    func didError(error: Error) {
        onResult(.withError(cause: error))
    }
}